packages/keyring-controller/src/KeyringController.ts

```diff
@@ -544,7 +751,30 @@ export class KeyringController extends BaseController<
       return this.getOrAddQRKeyring();
     }
 
-    return this.#keyring.addNewKeyring(type, opts);
+    const keyring = await this.#newKeyring(type, opts);
+
+    if (!keyring) {
+      throw new Error(KeyringControllerError.NoKeyring);
+    }
+
+    if (type === KeyringTypes.hd && (!isObject(opts) || !opts.mnemonic)) {
+      if (!keyring.generateRandomMnemonic) {
+        throw new Error(
+          KeyringControllerError.UnsupportedGenerateRandomMnemonic,
+        );
+      }
+
+      keyring.generateRandomMnemonic();
+      await keyring.addAccounts(1);
+    }
```

Comment:
I feel like this should be the responsibility of a hypothetical `init` method to be added on HDKeyring